src/main/kotlin/com/github/markoresic/footballvotingdemo/auth/AuthenticationService.kt

```diff
@@ -1,32 +1,106 @@
 package com.github.markoresic.footballvotingdemo.auth
 
 import com.github.markoresic.footballvotingdemo.config.JwtService
+import com.github.markoresic.footballvotingdemo.model.user.Role
 import com.github.markoresic.footballvotingdemo.model.user.User
-import com.github.markoresic.footballvotingdemo.repository.UserRepository
+import com.github.markoresic.footballvotingdemo.service.UserService
+import com.github.markoresic.footballvotingdemo.token.RefreshToken
+import com.github.markoresic.footballvotingdemo.token.RefreshTokenRepository
 import org.springframework.security.authentication.AuthenticationManager
+import org.springframework.security.authentication.BadCredentialsException
 import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
 import org.springframework.security.crypto.password.PasswordEncoder
 import org.springframework.stereotype.Service
 
+
 @Service
 class AuthenticationService(
-    private val userRepository: UserRepository,
-    private val passwordEncoder: PasswordEncoder,
+    private val userService: UserService,
+    private val refreshTokenRepository: RefreshTokenRepository,
     private val jwtService: JwtService,
     private val authenticationManager: AuthenticationManager
 ) {
 
     fun register(request: RegisterRequest): AuthenticationResponse {
-        val savedUser = userRepository.save(User(null, request.email, passwordEncoder.encode(request.password), request.nickname, request.role))
-        val jwtToken = jwtService.generateToken(savedUser)
-        return AuthenticationResponse(jwtToken)
+        val user = User(
+            null,
+            request.email,
+            passwordEncoder().encode(request.password),
+            request.nickname,
+            Role.USER
+        )
+        val savedUser = try {
+            userService.createUser(user)
+        } catch (e: Exception) {
+            throw RuntimeException("User with e-mail already exists!")
```

Comment:
This works now but It's very wired to trow RuntimeException here.  It's much safer not even try to save the user and check if email exists and return a custom exception to that. 

On top of that when you try to register a duplicate user you get generic 400 error with non-JSON body "User with e-mail already exists!" 

Please return json object as error and read this:

https://stackoverflow.com/questions/9269040/which-http-response-code-for-this-email-is-already-registered

and this:

https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-p2-semantics-18#section-7.4.10